src/components/TreeSelect.tsx

```diff
@@ -0,0 +1,168 @@
+import React, { useState, useRef } from "react";
+import Checkbox from "@leafygreen-ui/checkbox";
+import styled from "@emotion/styled";
+import { uiColors } from "@leafygreen-ui/palette";
+import { useOnClickOutside } from "hooks/useOnClickOutside";
+import Icon from "@leafygreen-ui/icon";
+
+interface Props {
+  state: string[];
+  tData: TreeDataEntry[];
+  onChange: (v: string[]) => void;
+  inputLabel: string;
+  optionsLabel: string;
+}
+export interface TreeDataEntry {
+  title: string;
+  value: string;
+  key: string;
+  children?: TreeDataEntry[];
+}
+
+export const TreeSelect = ({
+  state,
+  tData,
+  onChange,
+  inputLabel, // label for the select
+  optionsLabel // describes selected options
+}: Props) => {
+  const wrapperRef = useRef(null);
+  const [isVisible, setisVisible] = useState<boolean>(false);
+  useOnClickOutside(wrapperRef, () => setisVisible(false));
+  const toggleOptions = () => setisVisible(!isVisible);
+  return (
+    <Wrapper ref={wrapperRef}>
+      <BarWrapper onClick={toggleOptions}>
+        <LabelWrapper>
+          {inputLabel}
+          {optionsLabel}
+        </LabelWrapper>
+        <ArrowWrapper>
+          <div>
+            <Icon glyph={isVisible ? "ChevronUp" : "ChevronDown"} />
+          </div>
+        </ArrowWrapper>
+      </BarWrapper>
+      {isVisible && (
+        <OptionsWrapper>
+          {renderCheckboxes({ state, tData, onChange })}
+        </OptionsWrapper>
+      )}
+    </Wrapper>
+  );
+};
+
+// Executes when checkbox is clicked
+const handleOnChange = ({
+  state,
+  value,
+  onChange // callback function
+}: {
+  state: string[];
+  value: string;
+  onChange: (v: string[]) => void;
+}) => {
+  state.includes(value)
+    ? onChange(state.filter(v => v !== value))
+    : onChange([...state, value]);
+};
+
+// depth first traversal checkbox data.
+// pushes parent then children to rows array
+// keeps track of level for indentation
+const renderCheckboxes = ({
+  tData,
+  state,
+  onChange
+}: {
+  tData: TreeDataEntry[];
+  state: string[];
+  onChange: (v: [string]) => void;
+}) => {
+  const rows: JSX.Element[] = [];
+  let level = 0;
+  tData.forEach(entry => {
```

Comment:
This propagates the change handler for the entire tree even if you click on something other than the root. It seems to work so I'm probably missing something, but shouldn't you start on the node that's clicked? Possibly related, while testing I'm seeing that checking a parent doesn't check its children, but it should right?

Reply:
The checkboxes aren't nested so that's why there isn't any propagation. I added a `padding-left` based on the level. In the example, checking a parent doesn't check or uncheck it's children because I didn't handle it in the component that controls the state of TreeSelect.